File: app/src/main/java/org/fundacionparaguaya/advisorapp/viewmodels/EditPriorityViewModel.java
package org.fundacionparaguaya.advisorapp.viewmodels;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import org.fundacionparaguaya.advisorapp.models.IndicatorQuestion;
import org.fundacionparaguaya.advisorapp.repositories.SurveyRepository;
import org.joda.time.DateTime;
import org.joda.time.Months;

import java.util.Calendar;
import java.util.Date;

public final class EditPriorityViewModel extends ViewModel {
    private SurveyRepository repo;
    private LiveData<IndicatorQuestion> mIndicator = null;

    private String mReason;
    private String mAction;
    private Date mCompletionDate;

    EditPriorityViewModel(SurveyRepository repository) {
        this.repo = repository;
    }

    public final String getReason() {
        return this.mReason;
    }

    public final void setReason(String mReason) {
        this.mReason = mReason;
    }

    public final String getAction() {
        return this.mAction;
    }

    public final void setAction(String value) {
        this.mAction = value;
    }

    public final Date getCompletionDate() {
        return this.mCompletionDate;
    }

    public final void setCompletionDate(Date value) {
        this.mCompletionDate = value;
    }

    public final void setIndicator(int surveyId, int indicatorIndex) {
        mIndicator = Transformations.map(repo.getSurvey(surveyId), value -> value.getIndicatorQuestions().get(indicatorIndex));
    }

    public final LiveData<IndicatorQuestion> getIndicator() {
        if(mIndicator != null)
        {
            return mIndicator;
        }

        throw new IllegalStateException("getIndicator called before surveyId and IndicatorIndex set");
    }

    public void setNumMonths(int i) {
        Calendar cal = Calendar.getInstance();
        cal.add(Calendar.MONTH, i);
        this.mCompletionDate = cal.getTime();
    }

    public int getMonthsUntilCompletion()
    {
        DateTime start = new DateTime();
        DateTime end = new DateTime(mCompletionDate);

        return Months.monthsBetween(start, end).getMonths() + 1;
    }
}

File: app/src/main/java/org/fundacionparaguaya/advisorapp/viewmodels/EditPriorityViewModel.kt
package org.fundacionparaguaya.advisorapp.viewmodels

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel
import org.fundacionparaguaya.advisorapp.models.IndicatorQuestion
import org.fundacionparaguaya.advisorapp.models.Survey
import org.fundacionparaguaya.advisorapp.repositories.SurveyRepository

import java.util.Calendar
import java.util.Date


/**
 * View model used for the page that creates new/edits existing priorities.
 */

class EditPriorityViewModel(repository: SurveyRepository) : ViewModel() {

    private var repo: SurveyRepository? = null
    private var mSurvey: Survey? = null

    private var surveyId:Int? = null
    private var indicatorIndex:Int? = null

    var reason: String? = null
    var action: String? = null
    var completionDate: Date? = null


    init {
        repo = repository
    }

    fun setIndicator(surveyId: Int, indicatorIndex: Int)
    {
        this.surveyId = surveyId
        this.indicatorIndex = indicatorIndex
    }
    fun getIndicator(): LiveData<IndicatorQuestion> {

        surveyId?.let {
            indicatorIndex?.let {
                return Transformations.map(repo!!.getSurvey(surveyId!!)) { value -> value.indicatorQuestions[indicatorIndex!!] }
            }
        }

        throw Exception("getIndicator called before surveyId and IndicatorIndex set")
    }

    fun setNumMonths(i: Int) {
        val cal = Calendar.getInstance()
        cal.add(Calendar.MONTH, i)

        completionDate = cal.time
    }
}
